network-config-analyzer/MinimizeFWRules.py

```diff
@@ -0,0 +1,754 @@
+        else:
+            fw_rule = FWRule(grouped_elem, fixed_elem, self.connections)
+        return [fw_rule]
+
+    def _get_pods_grouping_by_labels_main(self, pods_set, extra_pods_set):
+        """
+        The main function to implement pods grouping by labels.
+        This function splits the pods into namespaces, anc per ns calls  get_pods_grouping_by_labels().
+        :param pods_set: the pods for grouping
+        :param extra_pods_set: additional pods that can be used for grouping
+        :return:
+        res_chosen_rep: a list of tuples (key,values,ns) -- as the chosen representation for grouping the pods.
+        res_remaining_pods: set of pods from pods_list that are not included in the grouping result (could not be grouped).
+        """
+        ns_context_options = set(pod.namespace for pod in pods_set)
+        res_chosen_rep = []
+        res_remaining_pods = set()
+        # grouping by pod-labels per each namespace separately
+        for ns in ns_context_options:
+            pods_set_per_ns = pods_set & self.cluster_info.ns_dict[ns]
+            extra_pods_set_per_ns = extra_pods_set & self.cluster_info.ns_dict[ns]
+            chosen_rep, remaining_pods = self._get_pods_grouping_by_labels(pods_set_per_ns, ns, extra_pods_set_per_ns)
+            res_chosen_rep.extend(chosen_rep)
+            res_remaining_pods |= remaining_pods
+        return res_chosen_rep, res_remaining_pods
+
+    def _get_pods_grouping_by_labels(self, pods_set, ns, extra_pods_set):
+        """
+        Implements pods grouping by labels in a single namespace.
+        :param pods_set: the set of pods for grouping.
+        :param ns: the namespace
+        :param extra_pods_set: additional pods that can be used for completing the grouping (originated in containing connections).
+        :return:
+        chosen_rep:  a list of tuples (key,values,ns) -- as the chosen representation for grouping the pods.
+        remaining_pods: set of pods from pods_list that are not included in the grouping result
+        """
+        if self.output_config.fwRulesDebug:
+            print('get_pods_grouping_by_labels:')
+            print('pods_list: ' + ','.join([str(pod) for pod in pods_set]))
+            print('extra_pods_list: ' + ','.join([str(pod) for pod in extra_pods_set]))
+        all_pods_set = pods_set | extra_pods_set
+        allowed_labels = self.cluster_info.allowed_labels
+        # labels_rep_options is a list of tuples (key, (values, pods-set)), where each tuple in this list is a valid
+        # grouping of pods-set by "key in values"
+        labels_rep_options = []
+        for key in allowed_labels:
+            values_for_key = self.cluster_info.get_values_set_for_key(key)
+            fully_covered_label_values = set()
+            pods_with_fully_covered_label_values = set()
+            for v in values_for_key:
+                all_pods_per_label_val = self.cluster_info.pods_labels_map[(key, v)] & self.cluster_info.ns_dict[ns]
```

Comment:
better use a variable to hold `self.cluster_info.ns_dict[ns]` and assign it outside the external loop

Reply:
done